r_code/synthetic_dgp.R

```diff
@@ -0,0 +1,1027 @@
+  stopifnot(rho < 1 & rho > -0.25)
+  stopifnot(rho * rho_shift < 1 & rho * rho_shift > -0.25)
+  # Require the intercept scale to be between -1 and 1.
+  stopifnot(intercept_scale <= 1 & intercept_scale >= -1)
+  # Require the intercept scale to be between -1 and 1.
+  stopifnot(loading_scale <= 1 & loading_scale >= -1)
+  # Stop if there are too few pre treat periods, 20 is a rule of thumb.
+  stopifnot(first_treat < 0.8 * num_periods | first_treat >= 20)
+
+
+  # Assign covariates and treatment given selection and overlap.
+  synth_data_unit <- .UnitLevelSimulation(
+    n_inp = num_entries,
+    type_inp = selection,
+    cov_overlap_inp = cov_overlap_scale,
+    loading_scale_inp = loading_scale,
+    num_factors_inp = num_factors,
+    int_scale_inp = intercept_scale,
+    rho_inp = rho,
+    rho_scale_inp = rho_scale,
+    rho_shift_inp = rho_shift,
+    prop_treated_inp = prop_treated,
+    treat_start = first_treat,
+    num_periods_inp = num_periods,
+    impact_mean_inp = treat_impact_mean,
+    impact_sd_inp = treat_impact_sd,
+    decay_mean_inp = treat_decay_mean,
+    decay_sd_inp = treat_decay_sd
+  )
+
+  # Work on time varying characteristics.
+  synth_data_factors <- .GenerateFactors(
+    num_factors_inp = num_factors,
+    num_periods_inp = num_periods,
+    freq_inp = freq
+  )
+  # Combine time and unit characteristics into a grid.
+  unit_time_grid <- tidyr::expand_grid(
+    entry = seq_len(num_entries),
+    time = seq_len(num_periods)
+  )
+
+  synth_data_full <- unit_time_grid %>%
+    dplyr::left_join(synth_data_unit, by = c("entry")) %>%
+    dplyr::left_join(synth_data_factors, by = c("time"))
+
+  rescale_y <- dplyr::case_when(
+    freq == "monthly" ~ log(rescale_y_mean),
+    freq == "weekly" ~ log(rescale_y_mean / 4),
+    freq == "daily" ~ log(rescale_y_mean / 30)
+  )
+  # Generate the counterfactual outcomes for each unit time combo.
+  synth_data_full <- .GenerateCounterfactual(synth_data_full,
+    num_periods_inp = num_periods,
+    rescale_y = rescale_y,
+    freq_inp = freq
+  )
+
+
+  # Generate the per period impact (taking account of decay).
+  synth_data_full <- .GenerateTreatImpact(
+    data_inp = synth_data_full,
+    cond_impact_inp =
+      conditional_impact_het
+  )
+
+  return(synth_data_full)
+}
+
+#' Draw IID noise and add to the outcome to finalize a draw from DGP.
+#'
+#' @param data_inp Long form Tibble, output of SyntheticDGP.
+#' @param seed Seed from which we draw noise, for reproducibility.
+#' @param log_output Boolean indicator for whether output should be log(y).
+#' @param sig_y Standard deviation of the iid normal (mean 0) noise.
+#'
+#' @return Noised long form tibble.
+# TODO(alexkellogg): Add non-iid noise process functionality.
+NoisifyDraw <- function(data_inp, seed, log_output = T, sig_y = 0.2) {
+  # Draws a vector or random normal (mean zero) epsilon values.
+  # This noise is added to the outcome, representing a new iid draw.
+  set.seed(seed)
```

Comment:
I think you're probably right. What I've done for now is to set the seed default value to NULL and skipped setting the seed if it's input is NULL. I want to think more about it, because a big part of writing papers evaluating methods requires the data to be comparable. I believe I could achieve this by resetting the seed before every call, but Ignacio has (in the past) included the seed as a parameter to randomize over the seed and compare methods. I've also seen this in other papers, so I want to have it (at least for now) just in case.